<a href="https://colab.research.google.com/github/Jonny-T87/Dojo-Work/blob/main/2nd_PCA_Exercise_(Core).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##PCA Exercise (Core)

Jonny Tesfahun
- 07/25/2022

In [7]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.pipeline import make_pipeline
from sklearn.neighbors import KNeighborsClassifier
from sklearn.decomposition import PCA
from sklearn.metrics import classification_report

In [8]:
# load the dataset
from sklearn.datasets import fetch_openml
mnist = fetch_openml('mnist_784')
# view the shape of the dataset
mnist.data.shape


(70000, 784)

In [9]:
mnist.target.shape

(70000,)

In [10]:
X = mnist.data
y = mnist.target

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

2. Prepare the Data:

In [11]:
#scaling data
scaler = StandardScaler()

#scaler.fit(X_train)
#X_train_scaled = scaler.fit_transform(X_train)

# using pca with 95% varience
pca = PCA(n_components=.95)

In [12]:
#using fit and transform on X train scaled data
#df_transformer_pca = pca.fit_transform(X_train_scaled)

3. Create 2 KNN models:

In [13]:
#Using knn and pca model
pca_knn_reg = KNeighborsClassifier()
pca_knn_reg.fit(X_train, y_train)
pca_knn_reg_pipe = make_pipeline(StandardScaler(), pca, pca_knn_reg)
pca_knn_reg_pipe.fit(X_train, y_train)

Pipeline(steps=[('standardscaler', StandardScaler()),
                ('pca', PCA(n_components=0.95)),
                ('kneighborsclassifier', KNeighborsClassifier())])

In [14]:
#Using no pca model
knn_reg = KNeighborsClassifier()
no_pca_knn_reg_pipe = make_pipeline(StandardScaler(), knn_reg)
no_pca_knn_reg_pipe.fit(X_train, y_train)

Pipeline(steps=[('standardscaler', StandardScaler()),
                ('kneighborsclassifier', KNeighborsClassifier())])

4. Evaluate and compare the models:

In [15]:
#Evaluating model with pca
%%time
preds_pca = pca_knn_reg_pipe.predict(X_test)

CPU times: user 56.7 s, sys: 2.97 s, total: 59.7 s
Wall time: 42.4 s


In [16]:
##Evaluating model without pca
%%time
preds_no_pca = no_pca_knn_reg_pipe.predict(X_test)

CPU times: user 1min 47s, sys: 1.96 s, total: 1min 49s
Wall time: 1min 7s


##Conclusion:

Definitely from both KNeighbors Classifier models, the model with the PCA ran faster in both CPU and Wall time. That model made the prediction in 42.4 seconds Wall time and 59.7 seconds CPU time. While the  KNeighbors Classifier model with no PCA, took 1 minute and 7 seconds Wall time and 1 minute 49 seconds CPU time to load. That is 24.6 seconds faster in Wall time and 50 seconds in CPU time. This is a big difference when considering that it had no hyper tuning or other manipulation of the data.